rfcs/20250321-graph-api-safe-softmax/README.md

```diff
@@ -0,0 +1,173 @@
+they can perform the check right before the division to minimize the performance
+impact:
+
+```math
+acc = sum(exp(x_0), ..., exp(x_s))
+```
+```math
+o_i = acc == 0 \; ? \; 0 \; : \; (exp(x_i) / acc)
+```
+
+With this option, we will have the same numeric behavior between oneDNN and
+PyTorch.
+
+One potential issue with this option is that the `-infinity`s may come from the
+attention score, rather than the attention mask. In this case, we will still
+return zeros for the rows with all `-infinity`s. It's possible that the user
+wants to propagate these special values to the output and the following layers.
+But it seems this is not considered in PyTorch's decision in [[#3]][3].
+
+### Option 3: Add an attribute to oneDNN Add/Select operations
+
+To address the potential issue in option 2, we will need to know where the
+`-infinity`s come from. This option proposes to add an attribute to the Add and
+Select operation which are used to apply the attention mask to the attention
+score as they know if the mask contains `-infinity`s. With the attribute being
+set, if the `-infinity` comes from the attention score side, the masked output
+will be `-infinity`, otherwise if the `-infinity` comes from the the attention
+mask side, a large negative value (eg. -1e9) will be applied the score instead.
+The following softmax layer will be kept unchanged. In this case, we will be
+able to propagate the special value to softmax's output if it comes from
+attention score.
+
+But with this option, the problem is for entire row with `-infinity` mask, if
+the large negative value is not large enough, softmax will generate the same
+probability (1.f / dim) for all elements. It's different from the zeros
+generated by safe softmax. If the large negative value is large enough, we will
+still see `NaN`s after a normal softmax, unless we change the implementation of
+softmax to detect these special values.
+
+```python
+
+>>> a = torch.tensor([-1e9, -1e9, -1e9, -1e9])
+>>> torch.nn.Softmax(dim=0)(a)
+tensor([0.2500, 0.2500, 0.2500, 0.2500])
+
+```
+
+### Conclusion
+
+Option 2 is more preferred based on the above discussion.
+
+## Primitive Proposal
+
+### Option 1: Introduce a new algorithm kind
+
+For a Graph SDPA implementation consisting of primitives to dispatch into a
+proper softmax implementation an additional marker on a softmax primitive side
+is required. While for the library side it is enough to have this marker not
+exposed in public API, for benchdnn validation it's necessary to have it
+publicly available to be able to dispatch into "inf_as_zero" implementation on
+the f32 path consisting of primitives. Otherwise, it will be impossible to
+achieve identical results.
+
+```c++
+/// Kinds of algorithms
+typedef enum {
+    ...
+    /// Softmax
+    dnnl_softmax_accurate = 0x30000,
+    /// Logsoftmax
+    dnnl_softmax_log,
+    /// Softmax treating all infinity inputs as zero
+    dnnl_softmax_inf_as_zero,
+} dnnl_alg_kind_t;
+```
```

Comment:
I agree with this primitive extension for validation purpose.
Though I am not sure we want to expose it publicly as we don't have plan to optimize that, it is only for fallback and validation.
Can we keep this alg_kind internal, for use by Graph API and validation only?

If we expose it externally, we will have to very clearly document it that from accuracy perspective, it should not be used outside of sdpa pattern, and is public only for validation purpose.

Reply:
Removed the algorithm from public headers, updated an RFC, and all related codes.